src/corecel/cont/detail/SpanImpl.hh

```diff
@@ -8,14 +8,50 @@
 #pragma once
 
 #include <cstddef>
+#include <type_traits>
 
 #include "corecel/Assert.hh"
 #include "corecel/Macros.hh"
+#include "corecel/OpaqueId.hh"
+#include "corecel/data/LdgIterator.hh"
 
 namespace celeritas
 {
 namespace detail
 {
+/*!
+ * Default type aliases for Span.
+ */
+template<class T>
+struct SpanTraits
+{
+    using element_type = T;
+    using pointer = std::add_pointer_t<element_type>;
+    using const_pointer = std::add_pointer_t<element_type const>;
+    using iterator = pointer;
+    using const_iterator = const_pointer;
+    using reference = std::add_lvalue_reference_t<element_type>;
+    using const_reference = std::add_lvalue_reference_t<element_type const>;
+};
+
+/*!
+ * Type aliases when data is read using __ldg. \c LdgValue checks that T is a
+ * valid type (const arithmetic or OpaqueId). Since \c LdgIterator returns a
+ * copy of the data read, we can't return a reference to the original data, we
+ * need to return a copy as well.
+ */
+template<class T>
+struct SpanTraits<LdgValue<T>>
+{
+    // element_type is always const
+    using element_type = typename LdgValue<T>::value_type;
```

Comment:
`element_type` should keep constness I think, it makes a difference in `Span(element_type (&arr)[N])` constructor. If we remove const here, then we can't construct a `LdgSpan` from a `T const[]`. Span can always remove const from the type if needed.

Reply:
Yeah I didn't realize that std array keeps the "const" in value type or that the double const didn't hurt. It gets away with it because it never returns value type.

Reply:
yes, Span has both element_type which keeps constness, and value_type which strips it, neither of them is ever returned so we just have to note that if we use them later. Maybe the naming of these aliases could be more clear.